src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java

```diff
@@ -365,46 +389,35 @@ private boolean checkSynchronized(final DetailAST synchronizedAst, boolean useBr
      * </pre>
      *
      * @param currentCase AST of the case that falls through to the next case.
-     * @param nextCase AST of the next case.
      * @return True if a relief comment was found
      */
-    private boolean hasFallThroughComment(DetailAST currentCase, DetailAST nextCase) {
-        boolean allThroughComment = false;
-        final int endLineNo = nextCase.getLineNo();
-
-        if (matchesComment(reliefPattern, endLineNo)) {
-            allThroughComment = true;
+    private boolean hasFallThroughComment(DetailAST currentCase) {
+        final DetailAST nextSibling = currentCase.getNextSibling();
+        final DetailAST ast;
+        if (nextSibling.getType() == TokenTypes.CASE_GROUP) {
+            ast = nextSibling.getFirstChild();
         }
         else {
-            final int startLineNo = currentCase.getLineNo();
-            for (int i = endLineNo - 2; i > startLineNo - 1; i--) {
-                final int[] line = getLineCodePoints(i);
-                if (!CodePointUtil.isBlank(line)) {
-                    allThroughComment = matchesComment(reliefPattern, i + 1);
-                    break;
-                }
-            }
+            ast = currentCase;
         }
-        return allThroughComment;
+        return hasReliefComment(ast);
     }
 
     /**
-     * Does a regular expression match on the given line and checks that a
-     * possible match is within a comment.
+     * Check if there is any fall through comment.
      *
-     * @param pattern The regular expression pattern to use.
-     * @param lineNo The line number in the file.
-     * @return True if a match was found inside a comment.
+     * @param ast ast to check
+     * @return true if relief comment found
      */
-    // suppress deprecation until https://github.com/checkstyle/checkstyle/issues/11166
-    @SuppressWarnings("deprecation")
-    private boolean matchesComment(Pattern pattern, int lineNo) {
-        final String line = getLine(lineNo - 1);
-
-        final Matcher matcher = pattern.matcher(line);
-        return matcher.find()
-                && getFileContents().hasIntersectionWithComment(
-                        lineNo, matcher.start(), lineNo, matcher.end());
+    private boolean hasReliefComment(DetailAST ast) {
+        boolean hasReliefComment = false;
+        final DetailAST nonCommentAst = getNextNonCommentAst(ast);
+        if (nonCommentAst != null) {
+            final DetailAST lastCommentAst = nonCommentAst.getPreviousSibling();
+            hasReliefComment = reliefPattern.matcher(lastCommentAst
+                    .getFirstChild().getText()).find();
+        }
+        return hasReliefComment;
```

Comment:
We can do this in a more succinct manner using optional:
```suggestion
        return Optional.ofNullable(getNextNonCommentAst(ast))
                .map(DetailAST::getPreviousSibling)
                .map(previous -> previous.getFirstChild().getText())
                .map(text -> reliefPattern.matcher(text).find())
                .orElse(false);
```

Reply:
Done

Reply:
@Kevin222004 For https://github.com/checkstyle/checkstyle/pull/12966#issuecomment-1694516053, we should be able to do Boolea.FALSE

Reply:
done